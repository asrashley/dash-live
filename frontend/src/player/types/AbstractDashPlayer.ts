import type { OptionMapWithChildren } from "@dashlive/options";
import { KeyParameters } from "./KeyParameters";

export interface DashPlayerProps {
    version?: string;
    autoplay?: boolean;
    videoElement: HTMLVideoElement;
    subtitlesElement?: HTMLDivElement;
    logEvent(eventName: string, text: string): void;
}
export abstract class AbstractDashPlayer {

    // eslint-disable-next-line no-useless-constructor
    constructor(protected props: DashPlayerProps) {
        /* no op */
    }

    abstract initialize(mpd: string, options: OptionMapWithChildren, keys: Map<string, KeyParameters>): Promise<void>;

    abstract destroy(): void;

    protected onCanPlayEvent = async () => {
        const { videoElement } = this.props;
        try {
            await videoElement.play();
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        } catch(err) {
            videoElement.muted = true;
            videoElement.play();
        }
        videoElement.removeEventListener('canplay', this.onCanPlayEvent);
    };

}