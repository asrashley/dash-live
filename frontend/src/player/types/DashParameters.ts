import type { OptionMapWithChildren } from "@dashlive/options";
import type { PeriodJson } from "./PeriodsJson";

export interface DashParameters {
    dash: {
        locationURL: string;
        mediaDuration: string;
        minBufferTime: string;
        mpd_id: string;
        now: string;
        periods: PeriodJson[];
        profiles: string[];
        publishTime: string;
        startNumber: number;
        suggestedPresentationDelay: number;
        timeSource: null;
        title: string;
    },
    options: OptionMapWithChildren;
    url: string;
}
