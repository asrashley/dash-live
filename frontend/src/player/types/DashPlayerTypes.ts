export type DashPlayerTypes = 'native' | 'shaka' | 'dashjs';