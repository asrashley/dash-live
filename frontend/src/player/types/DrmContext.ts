export type DrmContext = {
    laurl: string;
    scheme_id: string;
    version: number;
};
