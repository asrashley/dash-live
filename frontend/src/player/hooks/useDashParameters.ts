import { useCallback, useContext } from "preact/hooks";
import { useComputed, type ReadonlySignal } from "@preact/signals";

import { EndpointContext } from "../../endpoints";
import { useJsonRequest, UseJsonRequestHook } from "../../hooks/useJsonRequest";
import { DashParameters } from "../types/DashParameters";
import { KeyParameters } from "../types/KeyParameters";

interface UseDashParametersHook {
    dashParams: ReadonlySignal<DashParameters | undefined>;
    keys: ReadonlySignal<Map<string, KeyParameters>>;
    error: UseJsonRequestHook<DashParameters | undefined>['error'];
    loaded: UseJsonRequestHook<DashParameters | undefined>['loaded'];
}

export function useDashParameters(mode: string, stream: string, manifest: string, params: Readonly<URLSearchParams>): UseDashParametersHook {
    const apiRequests = useContext(EndpointContext);
    const request = useCallback((signal: AbortSignal) => apiRequests.getDashParameters(mode, stream, manifest, params,
        {
            signal,
        }), [apiRequests, manifest, mode, params, stream]);
    const { data: dashParams, loaded, error } = useJsonRequest<DashParameters | undefined>({
        request,
        initialData: undefined,
        name: 'dash parameters',
    });
    const keys = useComputed(() => {
        const rv = new Map();
        if (!dashParams.value?.dash) {
            return rv;
        }
        const { dash } = dashParams.value;
        for (const period of dash.periods) {
            for (const adaptation of period.adaptationSets) {
                for (const [kid, keypair] of Object.entries(adaptation.keys ?? {})) {
                    rv.set(kid, keypair);
                }
            }
        }
        return rv;
    });

    return { dashParams, keys, error, loaded };
}