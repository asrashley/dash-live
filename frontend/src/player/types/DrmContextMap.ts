import { DrmContext } from "./DrmContext";

export type DrmContextMap = {
    [drm: string]: DrmContext;
};