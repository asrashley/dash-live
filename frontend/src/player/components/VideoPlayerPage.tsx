import { useCallback, useContext, useEffect, useMemo, useRef } from "preact/hooks";
import { useSignal } from "@preact/signals";
import { useLocation, useParams } from "wouter-preact";

import { routeMap, uiRouteMap } from "@dashlive/routemap";

import { PlayerControls } from "../types/PlayerControls";
import { VideoPlayer } from "./VideoPlayer";
import { StatusEvent } from "../types/StatusEvent";
import { StatusPanel } from "./StatusPanel";
import { RouteParamsType } from "../../types/RouteParamsType";

import { DashPlayerTypes } from "../types/DashPlayerTypes";
import { useDashParameters } from "../hooks/useDashParameters";
import { LoadingSpinner } from "../../components/LoadingSpinner";

import { useSearchParams } from "../../hooks/useSearchParams";
import { AppStateContext } from "../../appState";

import "../styles/video.less";
import { PlaybackIconType } from "../types/PlaybackIconType";

export function manifestUrl(
  mode: string,
  stream: string,
  manifest: string,
  params: Readonly<URLSearchParams>
): string {
  const query = params.size ? `?${params.toString()}` : "";
  if (mode.startsWith("mps-")) {
    return `${routeMap.mpsManifest.url({
      mode: mode.slice(4),
      mps_name: stream,
      manifest,
    })}.mpd${query}`;
  }
  return `${routeMap.dashMpdV3.url({ mode, stream, manifest })}.mpd${query}`;
}

export interface KeyHandlerProps {
  controls: PlayerControls;
  setLocation: (url: string) => void;
  setIcon: (name: string) => void;
}

export function keyHandler(
  { controls, setIcon, setLocation }: KeyHandlerProps,
  ev: KeyboardEvent
) {
  switch (ev.key) {
    case " ":
    case "MediaPlayPause":
      if (controls.isPaused()) {
        setIcon("play");
        controls.play();
      } else {
        setIcon("pause");
        controls.pause();
      }
      break;
    case "Escape":
    case "MediaStop":
      setIcon("stop");
      controls.stop();
      break;
    case "MediaPlay":
      setIcon("play");
      controls.play();
      break;
    case "MediaPause":
      setIcon("pause");
      controls.pause();
      break;
    case "ArrowLeft":
    case "MediaTrackPrevious":
      setIcon("backward");
      controls.skip(-30);
      break;
    case "ArrowRight":
    case "MediaTrackNext":
      setIcon("forward");
      controls.skip(30);
      break;
    case "Home":
    case "Finish":
      setLocation(uiRouteMap.home.url());
      break;
  }
}

export default function VideoPlayerPage() {
  const [, setLocation] = useLocation();
  const { mode, stream, manifest } = useParams<RouteParamsType>();
  const { searchParams } = useSearchParams();
  const { cinemaMode } = useContext(AppStateContext);
  const currentTime = useSignal<number>(0);
  const controls = useSignal<PlayerControls | undefined>();
  const events = useSignal<StatusEvent[]>([]);
  const activeIcon = useSignal<PlaybackIconType | null>(null);
  const { dashParams, keys, loaded } = useDashParameters(
    mode,
    stream,
    manifest,
    searchParams
  );
  const iconTimer = useRef<number | undefined>();
  const mpd = useMemo<string>(
    () => manifestUrl(mode, stream, manifest, searchParams),
    [mode, stream, manifest, searchParams]
  );
  const playerName = useMemo<DashPlayerTypes>(() => {
    return (searchParams.get("player") ?? "native") as DashPlayerTypes;
  }, [searchParams]);
  const setIcon = useCallback((name: PlaybackIconType) => {
    activeIcon.value = name;
    window.clearTimeout(iconTimer.current);
    iconTimer.current = window.setTimeout(() => {
      activeIcon.value = null;
      iconTimer.current = undefined;
    }, 2000);
  }, [activeIcon]);
  const onKeyDown = useCallback((ev: KeyboardEvent) => {
    if (!controls.value) {
      return;
    }
    keyHandler({  controls: controls.value, setIcon, setLocation }, ev);
  }, [controls, setIcon, setLocation]);

  useEffect(() => {
    cinemaMode.value = true;
    return () => {
      cinemaMode.value = false;
    };
  }, [cinemaMode]);

  useEffect(() => {
    document.body.addEventListener("keydown", onKeyDown);
    return () => {
      document.body.removeEventListener("keydown", onKeyDown);
      window.clearTimeout(iconTimer.current);
      iconTimer.current = undefined;
    };
  }, [onKeyDown, iconTimer]);

  return (
    <div data-testid="video-player-page">
      {loaded.value ? (
        <VideoPlayer
          mpd={mpd}
          playerName={playerName}
          dashParams={dashParams}
          keys={keys}
          currentTime={currentTime}
          controls={controls}
          events={events}
          activeIcon={activeIcon}
        />
      ) : (
        <LoadingSpinner />
      )}
      <StatusPanel events={events} currentTime={currentTime} />
    </div>
  );
}
