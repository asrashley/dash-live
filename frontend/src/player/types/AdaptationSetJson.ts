import { DrmContextMap } from "./DrmContextMap";
import { KeyParameters } from "./KeyParameters";

export type AdaptationSetJson = {
    content_type: string;
    default_kid: string | null;
    drm?: DrmContextMap;
    fileSuffix: string;
    id: number;
    initURL: string;
    keys?: { [kid: string]: KeyParameters };
    lang: null | string;
    mediaURL: string;
    mimeType: string;
    mode: string;
    par: string;
    segmentAlignment: boolean;
    startWithSAP: number;
    timescale: number;
};
