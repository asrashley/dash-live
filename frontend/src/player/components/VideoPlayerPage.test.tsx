import {
  afterAll,
  afterEach,
  beforeEach,
  describe,
  expect,
  test,
  vi,
} from "vitest";
import { mock } from "vitest-mock-extended";
import fetchMock from "@fetch-mock/vitest";
import { useParams } from "wouter-preact";
import log from "loglevel";

import { useSearchParams } from "../../hooks/useSearchParams";
import { renderWithProviders } from "../../test/renderWithProviders";
import VideoPlayerPage, { manifestUrl } from "./VideoPlayerPage";
import { ApiRequests, EndpointContext } from "../../endpoints";
import dashParameters from "../../test/fixtures/play/vod/bbb/hand_made.json";
import { DashParameters } from "../types/DashParameters";

vi.mock("../../hooks/useSearchParams", () => ({
  useSearchParams: vi.fn(),
}));

vi.mock("wouter-preact", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useLocation: vi.fn(),
    useParams: vi.fn(),
  };
});

describe("VideoPlayerPage", () => {
  const apiRequests = mock<ApiRequests>();
  const mockUseSearchParams = vi.mocked(useSearchParams);
  const mockUseParams = vi.mocked(useParams);

  afterAll(() => {
    vi.restoreAllMocks();
  });

  beforeEach(() => {
    log.setLevel("error");
    apiRequests.getDashParameters.mockImplementation(
      async () => dashParameters as unknown as DashParameters
    );
    mockUseParams.mockReturnValue({
      mode: "dash",
      stream: "bbb",
      manifest: "hand-made",
    });
  });

  afterEach(async () => {
    vi.clearAllMocks();
    fetchMock.mockReset();
  });

  test("matches snapshot", async () => {
    const searchParams = new URLSearchParams({ player: "native" });
    mockUseSearchParams.mockReturnValue({
      searchParams,
    });
    const { asFragment, findBySelector, getByTestId } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <VideoPlayerPage />
      </EndpointContext.Provider>
    );
    getByTestId("video-player-page");
    await findBySelector("#vid-window");
    expect(asFragment()).toMatchSnapshot();
  });

  test("defaults to native player", async () => {
    const searchParams = new URLSearchParams();
    mockUseSearchParams.mockReturnValue({
      searchParams,
    });
    const { findBySelector } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <VideoPlayerPage />
      </EndpointContext.Provider>
    );
    await findBySelector("#vid-window");
  });

  test("shows loading spinner when loading", async () => {
    const searchParams = new URLSearchParams({ player: "native" });
    mockUseSearchParams.mockReturnValue({
      searchParams,
    });
    apiRequests.getDashParameters.mockClear();
    const { promise, resolve } = Promise.withResolvers<void>();
    apiRequests.getDashParameters.mockImplementation(async () => {
      await promise;
      return dashParameters as unknown as DashParameters;
    });
    const { getBySelector, findBySelector } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <VideoPlayerPage />
      </EndpointContext.Provider>
    );
    getBySelector(".lds-ring");
    resolve();
    await findBySelector("#vid-window");
  });

  test.each([
    ["vod", { player: "native" }, "/dash/vod/bbb/hand-made.mpd?player=native"],
    ["vod", { }, "/dash/vod/bbb/hand-made.mpd"],
    ["live", { }, "/dash/live/bbb/hand-made.mpd"],
    ["mps-vod", { player: "native" }, "/mps/vod/bbb/hand-made.mpd?player=native"],
    ["mps-vod", { }, "/mps/vod/bbb/hand-made.mpd"],
    ["mps-live", { }, "/mps/live/bbb/hand-made.mpd"],
  ])(
    "mode %s params %j generates manifest URL %s",
    (mode: string, params: Record<string, string>, expectedUrl: string) => {
      const searchParams = new URLSearchParams(params);
      expect(manifestUrl(mode, "bbb", "hand-made", searchParams)).toEqual(
        expectedUrl
      );
    }
  );
});
