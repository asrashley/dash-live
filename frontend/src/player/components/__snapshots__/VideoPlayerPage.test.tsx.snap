// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VideoPlayerPage > matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="video-player-page"
  >
    <div
      id="vid-window"
    >
      <span
        aria-label="pause icon"
        class="bi icon bi-pause fs-2"
        role="img"
      />
      <span
        aria-label="step-forward icon"
        class="bi icon bi-step-forward fs-2"
        role="img"
      />
      <span
        aria-label="step-backward icon"
        class="bi icon bi-step-backward fs-2"
        role="img"
      />
      <span
        aria-label="stop icon"
        class="bi icon bi-stop fs-2"
        role="img"
      />
      <video
        controls=""
        src="/dash/dash/bbb/hand-made.mpd?player=native"
      />
    </div>
    <div
      id="status"
    >
      <div
        class="play-position"
      >
        00:00:00
      </div>
    </div>
  </div>
</DocumentFragment>
`;
